### ISSUE TITLE
Continued issues with publishing modules to vsts package feed and manual nuget source configuration

### ISSUE THREAD

**Issue by whiggs** [PowerShell/PowerShellGet#245]
So I could explain the issues that I have been having with using the publish-module cmdlet to publish modules to a vsts package feed, and about how it was (somewhat) working in powershellget 1.5.0, but you broke it again in the latest version (see [this](https://adamrushuk.github.io/vsts/powershell/using-vsts-package-management-for-powershell-modules/) link), and how, even if I were using powershellget version 1.5.0, in order to successfully publish a module to a vsts package feed that does not have the nuget executable which powershellget uses (located at "C:\ProgramData\Microsoft\Windows\PowerShell\PowerShellGet\NuGet.exe") available on the system, we are forced to run publish-module twice (the first time to make the nuget exe available with the expectation the publish operation will fail and the second time after we manually add the source configuration to nuget which actually does publish the module), but why bother when there are numerous, numerous, numerous posts which have already reported this, and many of them quite old (over a year): #234, #133, #61, #101, #91, and lets go ahead and add this submission to the list.  So that makes a grand total of 6 requests to fix the issues with the powershellget as it pertains to being able to publish powershell modules to vsts feeds, as well as resolving the issues with nuget (having to manually add repo URL to source list), so my question is are there any plans to actually, you know, **FIX THE PROBLEM!?!?!?**  As I stated previously, around half of the listed issues were posted **OVER A YEAR AGO!!!**  The fact that this still remains a problem is UNACCEPTABLE!!!  I would humbly request that this issue be looked at and, if asking for a working, functional product isn't too much to ask of you, finally fixed.  Hopefully, you can break away from your busy nose picking schedule to actually work on the issue, because all I have seen when browsing through these issues is bug reports going un-answered by the developers (other users leave quite a few comments voicing their agreement) and left to rot.  There are many of us asking for resolutions to these issues, so, for the love of god, please actually fix the issues before we all die of old age.

**Comment by bmanikm** [comment]
@whiggs 
- Regarding the issue mentioned at [link](https://adamrushuk.github.io/vsts/powershell/using-vsts-package-management-for-powershell-modules/). `A parameter cannot be found that matches parameter name 'AllowPrereleaseVersions'` error should not happen when you launch a new PowerShell session. More details on this issue can be found at https://github.com/PowerShell/PowerShellGet/issues/208 . If you need to use the newly installed version of PowerShellGet within the current process, you will have to run `Import-Module -Name PowerShellGet -Force; Import-PackageProvider -Name PowerShellGet -Force`.

- Bootstrapping NuGet.exe to "C:\ProgramData\Microsoft\Windows\PowerShell\PowerShellGet\NuGet.exe"
```powershell
Invoke-RestMethod https://nuget.org/nuget.exe -OutFile "C:\ProgramData\Microsoft\Windows\PowerShell\PowerShellGet\NuGet.exe"
```
- For publishing to a VSTS feed, it is required to register the VSTS feed and credentials with NuGet.exe. NuGet.exe doesn't allow us to specify the credential as a parameter. Let us know if you have any ideas on resolving this.

- Apart from the enhancement issues, please let us know what other issues are blocking you from using the VSTS feed with PowerShellGet module. This will help us in prioritization.

**Comment by Remco-74** [comment]
@bmanikm : you ask for what are our blocking issues, but have you actually tried using it yourself and got it working? Using PowerShellGet with a VSTS / NuGet repository **is still a hot mess**.

- To start using it, you have to read a series of blogs or articles, all with varying levels or reliability.
- I used https://github.com/PowerShell/PowerShell-Docs/issues/1200#issue-227526581

- You still have to provide `-Credential`. Can those credentials not be stored somewhere secure? For example when using `Register-Repository`, you could use Export-Clixml to store the credential, readable for the current user only.

- Having to use `-Credential` blocks combining multiple repositories, as some need it, and some don't use it.

- `Find-Module` seems to work (somewhat), but `Publish-Module` fails all the time, and with errors that are not resolvable for a user. How about:

```
Publish-PSArtifactUtility : Failed to publish module 'My.Module': 'Cannot prompt for input in non-interactive mode.
'.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\1.6.0\PSModule.psm1:1447 char:17
+ ...             Publish-PSArtifactUtility @PublishPSArtifactUtility_Param ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : FailedToPublishTheModule,Publish-PSArtifactUtility
```
(using Module `PowerShellGet 1.6.0`, PackageProvider `PowerShellGet 1.6.0.0`, `NuGet 2.8.5.210`)

**Comment by bmanikm** [comment]
Yes, I have used VSTS repositories. I have also provided the required steps to @jianyunt to add the required documentation https://github.com/PowerShell/PowerShell-Docs/issues/1200.

Regarding credential persistence, please take a look at https://github.com/PowerShell/PowerShellGet/issues/105#issuecomment-369006689.

For publish-module error, did you specified the -Credential parameter for publishing to a VSTS feed as mentioned in https://github.com/PowerShell/PowerShell-Docs/issues/1200#issue-227526581 ?

**Comment by Remco-74** [comment]
@bmanikm yes, I am using the credential. See below:

```
PS C:\Users\My\PowerShell\Release> Get-PSRepository

Name                      InstallationPolicy   SourceLocation
----                      ------------------   --------------
PSGallery                 Untrusted            https://www.powershellgallery.com/api/v2/
company.PowerShell        Trusted              https://company.pkgs.visualstudio.com/_packaging/company.PowerShell/nuget/v2/


PS C:\Users\My\PowerShell\Release> Find-Module -Repository company.PowerShell -Credential $patCred

Version              Name                                Repository           Description
-------              ----                                ----------           -----------
1.0.0.0              company.PowerShell.Assert           company.PowerShell   company.PowerShell.Assert


PS C:\Users\My\PowerShell\Release> Publish-Module -Path .\company.PowerShell.Assert\ -NuGetApiKey company.PowerShell -Repository Goud
se.PowerShell -Credential $patCred
Publish-PSArtifactUtility : Failed to publish module 'company.PowerShell.Assert': 'Cannot prompt for input in non-interactive mode.
'.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\1.6.0\PSModule.psm1:1447 char:17
+ ...             Publish-PSArtifactUtility @PublishPSArtifactUtility_Param ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : FailedToPublishTheModule,Publish-PSArtifactUtility

PS C:\Users\My\PowerShell\Release> Get-Module PowerShellGet

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     1.6.0      PowerShellGet                       {Find-Command, Find-DscResource, Find-Module, Find-RoleCapability...}


PS C:\Users\My\PowerShell\Release> Get-PackageProvider

Name                     Version          DynamicOptions
----                     -------          --------------
msi                      3.0.0.0          AdditionalArguments
msu                      3.0.0.0
NuGet                    2.8.5.210        Destination, ExcludeVersion, Scope, SkipDependencies, Headers, FilterOnTag, Contains, AllowPrereleaseVersions, ConfigFile, SkipValidate
PowerShellGet            1.6.0.0          PackageManagementProvider, Type, Scope, AllowClobber, SkipPublisherCheck, InstallUpdate, NoPathUpdate, AllowPrereleaseVersions, Filter, Tag, Inc...
Programs                 3.0.0.0          IncludeWindowsInstaller, IncludeSystemComponent
```

**Comment by bmanikm** [comment]
As mentioned below at PowerShell/PowerShell-Docs#1200, for publishing to a VSTS feed, please register your source with NuGet.exe.

```powershell
# Publish modules to your VSTS package management feed using your PAT-base token.  
# For publishing, it is required to add your VSTS feed with username and PAT token, so that NuGet.exe uses these credential during push operation.
NuGet.exe sources Add -Name "psgettestfeed" -Source "https://myrepo.pkgs.visualstudio.com/_packaging/psgettestfeed/nuget/v2" -UserName manikb@microsoft.com -Password $myPatToken
```

For your VSTS feed, please run below with your user name and PAT as password.
```powershell
NuGet.exe sources Add -Name "https://company.pkgs.visualstudio.com/_packaging/company.PowerShell/nuget/v2/" -UserName $userName -Password $myPatToken
```
If NuGet.exe is not available on your machine, please refer to https://github.com/PowerShell/PowerShellGet/issues/245#issuecomment-373525017.

**Comment by Remco-74** [comment]
So, i finally got it working, It still isn't really as easy as it should be. 

- The double registration (`Register-PsRepository` and `nuget sources add`) is a weird requirement. A leaky metaphor at best, counterintuitive etc... There is much potential for things to go wrong here (mismatching registrations, etc..).

- Storing of Credentials is required (by both `PowerShellGet` and `nuget`), but the stored credentials don't actually allow users to actually skip `-Credentials`.

- The required use of `-Credential` doesn't work well when trying to find modules in multiple sources (combining public and private sources). 

    - You can either not specify it, and get two warnings for the missing `-Credential`, or 
    - You can specify it, but risk getting it intercepted by a public repository.
    - And never mind thinking about using multiple private repositories with differing credentials ..

- Getting it to work is error prone, labor intensive.

This process really needs improvement.
